[quant] Add reduce_range argument for qlinear_dynamic

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39187 [quant] Enable per-channel quantization for quantized nn.LSTM
* #39186 [quant] Make reduce_range default for RNNs
* #39125 [quant] Dynamic Linear module to use reduce_range
* **#39041 [quant] Add reduce_range argument for qlinear_dynamic**

Summary:
reduce_range option restricts the activation tensor to 7 bits instead of 8.
This is necessary to enable per channel quant for RNNs and LSTMs

Test Plan:
python test/test_quantization.py TestDynamicQuantizedLinear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21769691](https://our.internmc.facebook.com/intern/diff/D21769691)

supriyar committed:
[quant] Add reduce_range argument for qlinear_dynamic
Summary:
reduce_range option restricts the activation tensor to 7 bits instead of 8.
This is necessary to enable per channel quant for RNNs and LSTMs

Test Plan:
python test/test_quantization.py TestDynamicQuantizedLinear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp:
```diff
@@ -67,7 +67,9 @@ at::Tensor PackedLinearWeight::apply_dynamic_impl(at::Tensor input) {
-      /*preserve_sparsity=*/false);
+      /*preserve_sparsity=*/false,
+      /*force_scale_power_of_two=*/false,
+      /*reduce_range=*/reduce_range);
```
The logic in choose quantization params for reduce range is flaky. It works only for unsigned values: The logic currently is:
``` c++  
if (reduce_range) {
    qmin = 0;
    qmax = 127;
  }
we should have:
if (reduce_range)
 {
    qmin = qmin/2; // (0 ->0 and -128 -> -64)
    qmax = qmax/2; // Integer division (255->127 and 127 -> 63)
  }
```

raghuramank100 reviewed (approved):
We will also need to eventually remove the option of reduce_range in the aten ops. An example on how to do this is here: https://github.com/pytorch/pytorch/pull/35923/files. (The file check_backward_compatibility.py is relevant). In this PR, we create new aten ops for conv2d and conv2drelu and sunset the older ops down the road.  We need a way to remove the option of reduce_range in linear_dynamic and lstm ops down the road. (cc @jerryzh168 for more details)

supriyar committed:
Update on "[quant] Add reduce_range argument for qlinear_dynamic"


Summary:
reduce_range option restricts the activation tensor to 7 bits instead of 8.
This is necessary to enable per channel quant for RNNs and LSTMs

Test Plan:
python test/test_quantization.py TestDynamicQuantizedLinear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add reduce_range argument for qlinear_dynamic"


Summary:
reduce_range option restricts the activation tensor to 7 bits instead of 8.
This is necessary to enable per channel quant for RNNs and LSTMs

Test Plan:
python test/test_quantization.py TestDynamicQuantizedLinear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add reduce_range argument for qlinear_dynamic"


Summary:
reduce_range option restricts the activation tensor to 7 bits instead of 8.
This is necessary to enable per channel quant for RNNs and LSTMs

Test Plan:
python test/test_quantization.py TestDynamicQuantizedLinear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add reduce_range argument for qlinear_dynamic"


Summary:
reduce_range option restricts the activation tensor to 7 bits instead of 8.
This is necessary to enable per channel quant for RNNs and LSTMs

Test Plan:
python test/test_quantization.py TestDynamicQuantizedLinear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21769691](https://our.internmc.facebook.com/intern/diff/D21769691)

[ghstack-poisoned]
